### PR TITLE
Consider proposed text for draft-00

### PR DESCRIPTION
**Pull request by dkg**
We have an IETF draft filing deadline today.  I'm proposing submitting this as version -00.

I know it's a lot of churn, but i've been re-reading and thinking about overall document structure and i think it's an improvement.

### CODE REVIEW

File: draft-protected-headers.in.md [hunk 1]
```diff
@@ -59,7 +59,7 @@ normative:
  RFC8174:
 --- abstract
 
-This document describes a common strategy to extend the cryptographic protections provided by PGP/MIME, etc. to protect message headers in addition to message bodies.
+This document describes a common strategy to extend the end-to-end cryptographic protections provided by PGP/MIME, etc. to protect message headers in addition to message bodies.
```
Comment (juga0) [inline draft-protected-headers.in.md:62]: Hmm, something like [kuvert](https://github.com/az143/kuvert) could protect headers, but it's not end-to-end. Isn't the end-to-end part something transversal?

File: draft-protected-headers.in.md [hunk 2]
```diff
@@ -323,33 +327,32 @@ The revised algorithm for applying cryptographic protection to a message is as f
 Note that both new parameters, `obscured` and `legacy`, are effectively ignored if `crypto` does not contain encryption.
 This is by design, because they are irrelevant for signed-only cryptographic protections.
 
-Header Copying
---------------
+Copying All Headers
+-------------------
 
-All headers known to the MUA at composition time MUST be copied verbatim into the header of the target MIME part.
+All non-structural headers known to the composing MUA are copied to the MIME header of the Cryptographic Payload.
```
Comment (juga0) [inline draft-protected-headers.in.md]: should MIME header be defined or clarified?

File: draft-protected-headers.in.md [hunk 3]
```diff
@@ -75,7 +75,7 @@ This document defines a scheme, "Protected Headers for Cryptographic E-mail", wh
 
 This document describes how these protections can be applied to cryptographically signed messages, and also discusses some of the challenges of encrypting many transit-oriented headers.
 
-It offers guidance for encrypting non-transit-oriented headers like Subject, and also offers a means to preserve backwards compatibility so that an encrypted Subject remains available to recipients using software that does not implement support for the Protected Headers scheme.
+It offers guidance for protecting the confidentiality of non-transit-oriented headers like Subject, and also offers a means to preserve backwards compatibility so that an encrypted Subject remains available to recipients using software that does not implement support for the Protected Headers scheme.
```
Comment (juga0) [inline draft-protected-headers.in.md:78]: i understand what transit-header means, but is it defined somewhere?

File: draft-protected-headers.in.md [hunk 4]
```diff
@@ -323,33 +327,32 @@ The revised algorithm for applying cryptographic protection to a message is as f
...
-
-Encrypted Subject {#encrypted-subject}
------------------
+Confidential Subject {#confidential-subject}
+--------------------
 
 When a message is encrypted, the Subject should be obscured by replacing the Exposed Subject with three periods: `...`
 
-This value (`...`) was chosen because it is believed to be language agnostic and avoids communicating any potentially misleading information to the recipient (see {{common-pitfalls}} for a more detailed discussion).
+This value (`...`) was chosen because it is believed to be language agnostic and avoids communicating any potentially misleading information to the recipient (see {{misunderstood-obscured-subjects}} for a more detailed discussion).
```
Comment (juga0) [inline draft-protected-headers.in.md]: also in non-occidental languages?, but this is a different discussion...

File: draft-protected-headers.in.md [hunk 5]
```diff
@@ -100,7 +100,61 @@ For the purposes of this document, we define the following concepts:
...
+Of all the headers that an e-mail message may contain, only a handful are typically presented directly to the user.
+The user-facing headers are:
+
+ - `Subject`
+ - `From`
+ - `To`
+ - `Cc`
+ - `Date`
+
+The above is a complete list.  No other headers are considered "user-facing".
```
Comment (juga0) [inline draft-protected-headers.in.md:119]: Should all defined terms appear in ```` or with 1st letter capitalized?,
s/"user-facing"/`User-Facing`/

File: draft-protected-headers.in.md [hunk 6]
```diff
@@ -200,122 +254,120 @@ It is NOT RECOMMENDED to generate messages with such complicated structures.
...
-   2. When encrypting, Exposed Headers MAY be *obscured* by a transformation (including deletion).
-   3. When encrypting, if the MUA has obscured any user-facing header data, it SHOULD add a Legacy Display Part to the Cryptographic Payload which duplicates this information.
+We document legacy composition of cryptographically-protected messages (without protected headers) in {{legacy-composition}}, and then describe a revised version of that algorithm in {{protected-header-composition}} that produces conformant Protected Headers.
 
-*Note:* The above is not a description of a sequential algorithm.
+Copying All Headers
+-------------------
 
-Details:
+All non-structural headers known to the composing MUA are copied to the MIME header of the Cryptographic Payload.
```
Comment (juga0) [inline draft-protected-headers.in.md:274]: Should all defined terms appear in ```` or with 1st letter capitalized?, here:
s/All non-strucutural headers/All non-`Structural Headers`/

File: draft-protected-headers.in.md [hunk 7]
```diff
@@ -200,122 +254,120 @@ It is NOT RECOMMENDED to generate messages with such complicated structures.
...
+  - Set header `h` of `output` to `origheaders[h]`
+- Return `output`
 
-Aside from that limitation, this specification does not at this time define or limit the methods a MUA may use to convert Exposed Headers into Obscured Headers.
+Message Composition with Protected Headers {#protected-header-composition}
+------------------------------------------
 
-Legacy Display {#legacy-display}
-==============
+A reasonable sequential algorithm for composing a message *with* protected headers takes two more parameters in addition to `origbody`, `origheaders`, and `crypto`:
```
Comment (juga0) [inline draft-protected-headers.in.md:317]: s/reasonable//?

File: draft-protected-headers.in.md [hunk 8]
```diff
@@ -200,122 +254,120 @@ It is NOT RECOMMENDED to generate messages with such complicated structures.
...
+Legacy Display {#legacy-display}
+==============
 
-The above is a complete list.  No other headers are considered "typically visible".
+MUAs typically display user-facing headers ({{user-facing-headers}}) directly to the user.
+An encrypted message may be read by a decryption-capable legacy MUA that is unaware of this standard.
+The user of such a legacy client risks losing access to any obscured headers.
 
-Other headers may affect the visible rendering of the message (e.g., `References` and `In-Reply-To` may affect the placement of a message in a threaded discussion), but they are not directly displayed to the user and so are not considered "typically visible" for the purposes of this document.
+This section presents a workaround to mitigate this risk by restructuring the Cryptographic Payload before encrypting to include a "Legacy Display" part.
```
Comment (juga0) [inline draft-protected-headers.in.md:360]: s/"Legacy Display"/`Legacy Display`/?

**Comment by BjarniRunar** [comment]
I like this! :-)

There are a few FIXMEs still in there, and at least one unwritten section. I'm going to merge this PR into my own branch, make some quick fixes to those things and file another PR with the combination. You can then either merge or cherry-pick as you prefer?

**Review by juga0 — APPROVED** [review]
i went through the commits, then run `make` and read the resulting `html`. It makes sense to me. It's also great that now all the parts of the test vectors are generated automatically.
i just added some comments, mostly about the style in which the terms should appear. I've not gone through all the instances of those terms.

File: draft-protected-headers.in.md (hunk 2 above)
Comment (juga0) [reply draft-protected-headers.in.md]: I see now #15 :)

**Review by hpk42 — APPROVED** [review]
this looks good to me. thanks for pushing ahead!

File: draft-protected-headers.in.md (hunk 5 above)
Comment (dkg) [reply draft-protected-headers.in.md:119]: I've been trying to use `code` style to indicate literal textual elements, and Upper Case to indicate defined terms.

File: draft-protected-headers.in.md (hunk 1 above)
Comment (dkg) [reply draft-protected-headers.in.md:62]: I'm not sure what you mean by "transversal" but the point of this document is to try to focus on the PGP/MIME and S/MIME use cases, where cryptographic transformations are typically applied by the composing mail user agent. (e.g. as opposed to DKIM, which is designed for application by middleware transport).

Even if there are some systems (`kuvert` isn't the only one) that apply these transformations off of the client box, i think we can keep the "end-to-end" term as a distinguishing characteristic.

File: draft-protected-headers.in.md (hunk 4 above)
Comment (dkg) [reply draft-protected-headers.in.md]: hm, yes, this is a concern.  I've opened #19 to keep track of it.

File: draft-protected-headers.in.md (hunk 6 above)
Comment (dkg) [reply draft-protected-headers.in.md:274]: I've been trying to use `code` style to indicate literal textual elements, and Upper Case to indicate defined terms.

File: draft-protected-headers.in.md (hunk 8 above)
Comment (dkg) [reply draft-protected-headers.in.md:360]: I've been trying to use `code` style to indicate literal textual elements, and Upper Case to indicate defined terms.

File: draft-protected-headers.in.md (hunk 3 above)
Comment (dkg) [reply draft-protected-headers.in.md:78]: I've opened #20 about this.

**Pull request merged**